sycl/doc/Assert.md

```diff
@@ -0,0 +1,421 @@
+   2. Low-Level Runtime prints assert failure message to `stderr`
+   3. Low-Level Runtime calls `abort()`
+
+
+## Fallback approach
+
+If Device-side Runtime doesn't support `__devicelib_assert_fail` then a fallback
+approach comes in place. The approach doesn't require any support from
+Device-side Runtime and Native Device Compiler. Neither it does from Low-level
+Runtime.
+
+Within this approach, a mutable program scope variable is introduced. This
+variable stores a flag which says if an assert failure was encountered. Fallback
+implementation of `__devicelib_assert_fail` atomically raises the flag so that
+DPCPP Runtime is able to detect assert failure after kernel finishes.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+   2. Fallback version of `__devicelib_assert_fail` is called
+   3. Assert information is stored into program-scope variable
+   4. Kernel continues running
+ - Host side:
+   1. A copy 'kernel' is enqueued as the one depending on user's kernel to get
+      the value of assert failure flag.
+   2. A host-task is enqueued to check value of assert failure flag.
+   3. The host task calls abort whenever assert failure flag is set.
+
+Illustrating this with an example, lets assume the user enqueues three kernels:
+ - `Kernel #1`
+ - `Kernel #2`
+ - `Kernel #3`, which depends on `Kernel #1`
+
+The resulting graph will look like this: ![graph](images/assert-fallback-graph.svg)
+
+### Interface to program scope variable
+
+Multiple translation units could be compiled/linked into a single device binary
+image. All of them should have `extern` declaration of program scope variable
+available. Definition of the variable is only available within devicelib in the
+same binary image where fallback `__devicelib_assert_fail` resides.
+
+<a name="prog-scope-var-decl">The variable has the following structure and
+declaration:</a>
+
+```c++
+struct AssertHappened {
+  int Flag = 0;
+};
+
+#ifdef __SYCL_DEVICE_ONLY__
+extern SYCL_GLOBAL_VAR AssertHappened AssertHappenedMem;
+#endif
+```
+
+Here, `SYCL_GLOBAL_VAR` is a macro which wraps special attribute to allow for
+mutable program-scope variable.
+
+The reference to extern variable is resolved within online-linking against
+fallback devicelib.
+
+### Online-linking fallback `__devicelib_assert_fail`
+
+Online linking against fallback implementation of `__devicelib_assert_fail` is
+performed only when assertion is enabled and Device-side Runtime doesn't provide
+implementation of `__devicelib_assert_fail`.
+
+In DPCPP headers one can see if assert is enabled with status of `NDEBUG` macro
+with `#ifdef`'s. This allows to enqueue a copy kernel and host task. The copy
+kernel will copy `AssertHappenedMem` to host and host-task will check the `Flag`
+value and `abort()` as needed.
+
+When in DPCPP Runtime Library this knowledge is obtained from device binary
+image descriptor's property sets.
```

Comment:
Not clear if the enqueueing of these auxiliary kernels happens from under #ifndef NDEBUG, or after checking the property. Please clarify here.

Reply:
Clarified.